src/server/pkg/storage/chunker/chunker.go

```diff
@@ -0,0 +1,69 @@
+package chunker
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/chmduquesne/rollinghash/buzhash64"
+)
+
+const (
+	// MB is Megabytes.
+	MB = 1024 * 1024
+	// MinSize is the minimum chunk size.
+	MinSize = 5 * MB
+	// AverageBits determines the average chunk size (2^AverageBits).
+	AverageBits = 23
+	// MaxSize is the maximum chunk size.
+	MaxSize = 15 * MB
+	// WindowSize is the size of the rolling hash window.
+	WindowSize = 64
+)
+
+// Chunker is a content defined chunker.
+// Chunk split points are determined by a bit pattern in a rolling hash function (buzhash64 at https://github.com/chmduquesne/rollinghash).
+// A min/max chunk size is also enforced to avoid creating chunks that are too small/big.
+type Chunker struct {
+	f         func(io.Reader) error
+	buf       *bytes.Buffer
+	hash      *buzhash64.Buzhash64
+	splitMask uint64
+}
```

Comment:
I think this interface will work out better in the long run if we make take a `Reader`, rather than implementing `Writer`, because it means that you'll have direct control over the flow of data from inside the implementation, rather than having to wait for data to be written. It seems like `bufio.Scanner` should be a pretty good source of inspiration for this, since it's doing something similar. In fact it might make sense to just do a simple implementation by implementing a `SplitFunc` and using a `bufio.Scanner` for this and then come back to it later if that breaks down.

Reply:
I don't think this would make sense for how we would be using this. I think taking a Reader makes sense if you are pulling from a source of existing data (a file or memory buffer), but we are actually going to chain this at the end of a streaming write/merge process which will be writing out data as it generates it. If we took a reader, we would need to setup a pipe to convert a write to a read just for the purpose of having a reader for the chunker. The "direct control over the flow of data" idea does not seem to apply if the data does not exist until it is generated.